roles/forklift_versions/lookup_plugins/forklift_versions.py

```diff
@@ -0,0 +1,94 @@
+          - this lookup will match the katello version for all other scenarios
+"""
+import yaml
+from ansible.errors import AnsibleParserError, AnsibleLookupError
+from ansible.plugins.lookup import LookupBase
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+def version_sort_key(version):
+    try:
+        return [int(u) for u in version.split('.')]
+    except ValueError:
+        return [9999, 9999, 9999]
+
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables=None, **kwargs):
+        ret = []
+
+        for term in terms:
+            display.debug("Forklift lookup term: %s" % term)
+
+            lookup_params = dict(x.split('=', 1) for x in term.split())
+
+            try:
+                scenario = lookup_params['scenario']
+                if scenario != 'foreman':
+                    scenario = 'katello'
+                scenario_version = lookup_params['scenario_version']
+                versions_file_name = lookup_params['file']
+                upgrade = lookup_params.get('upgrade', False)
+            except KeyError:
+                raise AnsibleParserError("missing required param for forklift_version")
+
+            try:
+                with open(versions_file_name, 'r') as versions_file:
+                    versions = yaml.safe_load(versions_file)
+            except Exception:
+                raise AnsibleLookupError("couldn't read '%s'" % versions_file_name)
+
+            if not upgrade:
+                for version in versions['installers']:
+                    if version[scenario] == scenario_version:
+                        forklift_vars = {
+                                'foreman_repositories_version': version['foreman'],
+                                'foreman_client_repositories_version': version['foreman'],
+                                'katello_repositories_version': version['katello'],
+                                'katello_repositories_pulp_version': version['pulp'],
+                                'pulp_repositories_version': version['pulp'],
+                                'puppet_repositories_version': version['puppet'],
+                                }
+                        ret.append(forklift_vars)
+                        break
+            else:
+                upgrade_versions = set()
+                for version in reversed(versions['installers']):
+                    if version[scenario] == scenario_version:
+                        upgrade_versions.add(scenario_version)
+                    elif 1 <= len(upgrade_versions) < 3:
+                        upgrade_versions.add(version[scenario])
+                upgrade_versions = sorted(upgrade_versions, key=version_sort_key)
+                while len(upgrade_versions) < 3:
+                    upgrade_versions.insert(0, upgrade_versions[0])
```

Comment:
Catch an `IndexError` in case the specified version wasn't in the versions file at all?

I was also thinking about how clear this is, but it's probably less readable:

```python
upgrade_versions = [upgrade_versions[0]] * (3 - len(upgrade_versions)) + upgrade_versions
```

Should 3 be a constant somewhere?

Reply:
I don't think that's readable ;-)

Implemented the two other suggestions